Add mixin class to replace mandatory hooks

casparvl opened the pull request:
This no longer depends on https://github.com/EESSI/test-suite/pull/185
We can add the logging of the version back in as a part of https://github.com/EESSI/test-suite/pull/185

smoors reviewed (commented):
thanks a lot for this initial effort toward a mixin class, and for identifying possible issues!

smoors commented on eessi/testsuite/eessi_mixin.py:
```diff
@@ -0,0 +1,118 @@
+    - Init phase: device_type, scale, module_name
+    - Setup phase: compute_unit, required_mem_per_node
+    """
+
```
we can define here the parameters that are required as "abstract parameters", e.g.
```
 module_name = parameter()
```
which ensures that the parameter is always set in the test.

casparvl replied:
Yes, but the problem is that if the child class doesn't overwrite that, the error is _extremely_ unclear:

```
WARNING: skipping test 'EESSI_LAMMPS_lj': test has one or more undefined parameters
WARNING: skipping test 'EESSI_LAMMPS_rhodo': test has one or more undefined parameters
```
It might get better if this issue https://github.com/reframe-hpc/reframe/issues/3254 gets resolved. In that case, this approach is indeed much cleaner. The issue is on the backlog for the ReFrame 4.7 milestone, and I gave them a piece of code that _should_ do the job. So there is a fair chance we'll get it. In which case your approach is indeed cleaner.

The one downside (and this is a major one, since it is not trivial) is that we then cannot state _when_ we need that parameter. Some of these, the user might only be able to set after the `setup` phase, because they want to use information from the `current_partition`. They need to then know if that is 'early enough' for what the Mixin class requires (i.e. when the hooks called in the Mixin class require that variable to be set).

smoors replied:
stating which parameter is undefined would indeed be a welcome improvement.

however, `parameter`s should always be defined in the class body, there is no doubt when to set it.
see also the discussion in slack: https://reframetalk.slack.com/archives/CF8ENLUDB/p1725607966946879

> For now, I'd like to go for the verbose option of raising the error ourselves, so that we control what it says.

i'm fine with that, i also think an error is better than a warning here. so this should be done in the class body of the mixin class, right?

for normal or user-defined variables, we'll probably have to roll our own in any case indeed.

casparvl commented on eessi/testsuite/tests/apps/lammps/lammps.py:
```diff
@@ -8,18 +8,26 @@
+    # scales = list(SCALES.keys())
+    # scales.append('blabla')
+    # scale = parameter(scales)
     scale = parameter(SCALES.keys())
     valid_prog_environs = ['default']
     valid_systems = ['*']
     time_limit = '30m'
```
I've just tried to move that, but it doesn't work for `valid_prog_environs`, `valid_systems` and `time_limit`. I'm getting

```
WARNING: skipping test file '/gpfs/home4/casparl/EESSI/test-suite/eessi/testsuite/tests/apps/lammps/lammps.py': reframe syntax error: 'time_limit' already defined in class 'EESSI_LAMMPS_base' (rerun with '-v' for more information)
```
even though I _did_ remove them from the lammp class itself.

smoors replied:
that's unfortunate, but luckily there's an easy workaround, for example adding this to the mixin class works:

```py
    @run_after('init')
    def set_time_limit(self):
        if not self.time_limit:
            self.time_limit = '1h'
```

casparvl commented:
Hm, I went down the rabbit hole of not trying to use hardcoded versions. And then also being able to use them at runtime. It works well from a git checkout. However if I
```
pip install git+https://github.com/casparvl/test-suite.git@use_mixin_class
```
The version that is used for some reason is 0.1.dev753+g97e60ab. Although I'm not sure how it is supposed to get any other version: for git, it can check the tags and increment by one. Where should it get it's version (or previous version) if you pip-install out of the blue?

casparvl commented:
I was being silly. The `use_mixin_class` is a branch on _my_ fork. Tags are not forked, and thus my fork did not contain _any_ tags.

I now added a tag `v0.3.3` to the latest commit in my `use_mixin_class` branch. Then:

```
(my_test_venv) [casparl@tcn1 test-suite]$ pip install git+https://github.com/casparvl/test-suite.git@use_mixin_class
Collecting git+https://github.com/casparvl/test-suite.git@use_mixin_class
  Cloning https://github.com/casparvl/test-suite.git (to revision use_mixin_class) to /tmp/pip-req-build-g9hmxril
  Running command git clone -q https://github.com/casparvl/test-suite.git /tmp/pip-req-build-g9hmxril
  Running command git checkout -b use_mixin_class --track origin/use_mixin_class
  Switched to a new branch 'use_mixin_class'
  branch 'use_mixin_class' set up to track 'origin/use_mixin_class'.
  Resolved https://github.com/casparvl/test-suite.git to commit 97e60abb37126bb7165ec66197459f0b08320f88
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Building wheels for collected packages: eessi-testsuite
  Building wheel for eessi-testsuite (PEP 517) ... done
  Created wheel for eessi-testsuite: filename=eessi_testsuite-0.3.3-py3-none-any.whl size=1854649 sha256=ff462274cc8995e5e592378f90a0c05c65a26dbbc7f441da9a0ea373d11ee0de
  Stored in directory: /tmp/pip-ephem-wheel-cache-k8t31ry2/wheels/65/6f/b7/6fb6ee1323b931899a049e67caaadafd3ee34eb1b811ca39ae
Successfully built eessi-testsuite
Installing collected packages: eessi-testsuite
Successfully installed eessi-testsuite-0.3.3
```
As expected.

casparvl commented:
I've also followed the steps for release documented in our own readme, to see if that would also give us the right versioning. It does:
```
$ python setup.py sdist
...
$ ls -al dist/eessi-testsuite-0.3.3.tar.gz
-rw-r----- 1 casparl casparl 1844717 Sep 20 16:34 dist/eessi-testsuite-0.3.3.tar.gz
```
Awesome, I think that automatic versioning is working perfectly fine!

casparvl commented:
Oh, we should probably list `setuptools_scm` as an (optional?) dependency somewhere, since that will be needed to retreive the version at runtime.

Maybe as an improvement, I should also configure our setup procedure to create a version file, and read from there at runtime as the _first_ option. At least when it is an _installed_ package, that should work. If it's just a git clone, one would still need `setuptools_scm` to do the runtime version check.

boegel commented:
> Oh, we should probably list `setuptools_scm` as an (optional?) dependency somewhere, since that will be needed to retreive the version at runtime.

I have some bad experiences with `setuptools_scm`, if we can avoid it I would stay away from it tbh...

Unless it's a huge win somehow and makes our lives a lot easier.
It may seem to be working out fine, but I've seen it act up once too many times to "trust" it.

Then again, maybe I cry too easily when managing packages... ;)

casparvl commented:
Well, documentation of `setuptools_scm` isn't the best, but with a bit of fiddling, I've actually found it does what I want in all scenarios I could think of.

Sam suggested to me that the `eessi_mixin` class was a nice opportunity to make sure all tests write the current version of the test suite as a loggable variable. This triggered me to look at how to best make sure I could import a version, without duplicating the versions in `setup.py`, `pyproject.toml` and friends. `setuptools_scm` allows me to do that _and more_. It also means that if we run the test suite from unrelease commits, the logged version reflects that (and even reflects if the working dir is 'dirty' with uncommitted changes). In other words: from the ReFrame report file, I'd be able to infer pretty much exactly what was run.

This is a great benefit when trying to interpret sudden changes in test performance: if we see the versions of those test runs were not equal, that provides the first (possible) explaination.

Also, I'm lazy and forgetful, so not having to update the versions upon release is something I find very nice :)

If it causes trouble, we can always go back to hardcoded version (but paying attention we only put it in one place). And, remember that we need `setuptools_scm` to build the package distribution, but it is _not_ a dependency for running the test suite (_unless_ you are running it from a git-clone. And even then, you _can_ run it without, you'll just get a `0.0.0` version logged...)

casparvl commented:
Ok, it is a bit silly, since the problem I have been fixing is that we can _build_ the distribution on all python versions in our CI. I don't think that's super important, we really want to be able to _install_ on all version in our CI. Anyway, it works everywhere now. It required pushing the requirement for `setuptools_scm` to `setup.py`, because we require logic, as versions will depend on the python version. There is no single version that works for all.

casparvl commented:
I've split of the whole `setuptools_scm` versioning logic into a separate PR https://github.com/EESSI/test-suite/pull/185 

I'd like to get that one merged for the new release (0.3.3). This mixin class will probably take more time, and I don't want to wait with releasing 0.3.3 for this to be finished.

smoors commented on eessi/testsuite/tests/apps/lammps/lammps.py:
```diff
@@ -8,18 +8,23 @@
     valid_prog_environs = ['default']
     valid_systems = ['*']
     time_limit = '30m'
+
```
we can use an easy workaround to move those 3 variables into the mixin class, for example for time_limit:

```py
    @run_after('init')
    def set_time_limit(self):
        if not self.time_limit:
            self.time_limit = '1h'
```

casparvl replied:
Great, thanks Sam for this clever solution!

smoors reviewed (approved):
works like a charm, this will help a lot getting more people to write testing!
